File: VB/HyperLinkPropertyEditor.Win/WinHyperLinkStringPropertyEditor.vb
Imports System
Imports DevExpress.ExpressApp
Imports DevExpress.XtraEditors
Imports DevExpress.ExpressApp.Model
Imports DevExpress.XtraEditors.Mask
Imports DevExpress.ExpressApp.Editors
Imports System.Text.RegularExpressions
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.ExpressApp.Win.Editors
Imports DevExpress.XtraEditors.Repository

Namespace HyperLinkPropertyEditor.Win
    <PropertyEditor(GetType(System.String), "HyperLinkStringPropertyEditor", False)>
    Public Class WinHyperLinkStringPropertyEditor
        Inherits StringPropertyEditor

        'Dennis TODO: This is to be setup via the Model Editor at the ViewItems | PropertyEditors | HyperLinkStringPropertyEditor level once.
        Public Const UrlEmailMask As String = "(((http|https|ftp)\://)?[a-zA-Z0-9\-\.]+\.[a-zA-Z]{2,3}(:[a-zA-Z0-9]*)?/?([a-zA-Z0-9\-\._\?\,\'/\\\+&amp;amp;%\$#\=~])*)|([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,6})"
        Private hyperlinkEditCore As HyperLinkEdit
        Public Sub New(ByVal objectType As Type, ByVal info As IModelMemberViewItem)
            MyBase.New(objectType, info)
        End Sub
        Public Shadows ReadOnly Property Control() As HyperLinkEdit
            Get
                Return hyperlinkEditCore
            End Get
        End Property
        Protected Overrides Function CreateRepositoryItem() As RepositoryItem
            Return New RepositoryItemHyperLinkEdit()
        End Function
        Protected Overrides Function CreateControlCore() As Object
            hyperlinkEditCore = New HyperLinkEdit()
            Return hyperlinkEditCore
        End Function
        Protected Overrides Sub SetupRepositoryItem(ByVal item As RepositoryItem)
            MyBase.SetupRepositoryItem(item)
            Dim hyperLinkProperties As RepositoryItemHyperLinkEdit = CType(item, RepositoryItemHyperLinkEdit)
            hyperLinkProperties.SingleClick = TypeOf View Is ListView
            hyperLinkProperties.TextEditStyle = TextEditStyles.Standard
            AddHandler hyperLinkProperties.OpenLink, AddressOf hyperLinkProperties_OpenLink
            EditMaskType = EditMaskType.RegEx
            hyperLinkProperties.Mask.MaskType = MaskType.RegEx
            hyperLinkProperties.Mask.EditMask = UrlEmailMask
        End Sub
        Private Sub hyperLinkProperties_OpenLink(ByVal sender As Object, ByVal e As OpenLinkEventArgs)
            e.EditValue = GetResolvedUrl(e.EditValue)
        End Sub
        Public Shared Function GetResolvedUrl(ByVal value As Object) As String
            Dim url As String = Convert.ToString(value)
            If Not String.IsNullOrEmpty(url) Then
                If url.Contains("@") AndAlso IsValidUrl(url) Then
                    Return String.Format("mailto:{0}", url)
                End If
                If Not url.Contains("://") Then
                    url = String.Format("http://{0}", url)
                End If
                If IsValidUrl(url) Then
                    Return url
                End If
            End If
            Return String.Empty
        End Function
        Private Shared Function IsValidUrl(ByVal url As String) As Boolean
            Return Regex.IsMatch(url, UrlEmailMask)
        End Function
    End Class
End Namespace

File: CS/XPO/HyperLinkEditor/HyperLinkEditor.Win/Editors/WinHyperLinkStringPropertyEditor.cs
using System;
using DevExpress.ExpressApp;
using DevExpress.XtraEditors;
using DevExpress.ExpressApp.Model;
using DevExpress.XtraEditors.Mask;
using DevExpress.ExpressApp.Editors;
using System.Text.RegularExpressions;
using DevExpress.XtraEditors.Controls;
using DevExpress.ExpressApp.Win.Editors;
using DevExpress.XtraEditors.Repository;

namespace HyperLinkPropertyEditor.Win {
	[PropertyEditor(typeof(System.String),  false)]
    public class WinHyperLinkStringPropertyEditor : StringPropertyEditor {
        public const string UrlEmailMask = @"(((http|https|ftp)\://)?[a-zA-Z0-9\-\.]+\.[a-zA-Z]{2,3}(:[a-zA-Z0-9]*)?/?([a-zA-Z0-9\-\._\?\,\'/\\\+&amp;amp;%\$#\=~])*)|([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,6})";
        private HyperLinkEdit hyperlinkEditCore;
        public WinHyperLinkStringPropertyEditor(Type objectType, IModelMemberViewItem info)
            : base(objectType, info) {
        }
        public new HyperLinkEdit Control { get { return hyperlinkEditCore; } }
        protected override RepositoryItem CreateRepositoryItem() {
            return new RepositoryItemHyperLinkEdit();
        }
        protected override object CreateControlCore() {
            hyperlinkEditCore = new HyperLinkEdit();
            return hyperlinkEditCore;
        }
        protected override void SetupRepositoryItem(RepositoryItem item) {
            base.SetupRepositoryItem(item);
            RepositoryItemHyperLinkEdit hyperLinkProperties = (RepositoryItemHyperLinkEdit)item;
            hyperLinkProperties.SingleClick = View is ListView;
            hyperLinkProperties.TextEditStyle = TextEditStyles.Standard;
            hyperLinkProperties.OpenLink += hyperLinkProperties_OpenLink;
            EditMaskType = EditMaskType.RegEx;
            hyperLinkProperties.Mask.MaskType = MaskType.RegEx;
            hyperLinkProperties.Mask.EditMask = UrlEmailMask;
        }
        private void hyperLinkProperties_OpenLink(object sender, OpenLinkEventArgs e) {
            e.EditValue = GetResolvedUrl(e.EditValue);
        }
        public static string GetResolvedUrl(object value) {
            string url = Convert.ToString(value);
            if (!string.IsNullOrEmpty(url)) {
                if (url.Contains("@") && IsValidUrl(url))
                    return string.Format("mailto:{0}", url);
                if (!url.Contains("://"))
                    url = string.Format("http://{0}", url);
                if (IsValidUrl(url))
                    return url;
            }
            return string.Empty;
        }
        private static bool IsValidUrl(string url) {
            return Regex.IsMatch(url, UrlEmailMask);
        }
    }
}